src/io/ioexternalbus.cpp

```diff
@@ -28,10 +50,217 @@ QString IOExternalBus::description() const {
 }
 
 VInt IOExternalBus::ioRead(AInt offset, unsigned size) {
-    return 0;
+    uint32_t rvalue = 0;
+
+    if (tcpSocket->isOpen()) {
+        while (skt_use)
+            QApplication::processEvents();  // TODO only need in rvss, change to mutex
+        skt_use = 1;
+        uint32_t payload = htonl(offset);
+        QByteArray dp = QByteArray(static_cast<const char*>((void*)&payload), 4);
+        if (send_cmd(VB_PREAD, 4, &dp) < 0) {
+            skt_use = 0;
+            return 0;
+        }
+
+        cmd_header_t cmd_header;
+
+        if (recv_cmd(cmd_header) < 0) {
+            skt_use = 0;
+            return 0;
+        }
+
+        if (cmd_header.payload_size) {
+            dp = QByteArray(8, 0);
+            if (recv_payload(dp, 8) < 0) {
+                skt_use = 0;
+                return 0;
+            }
+            uint32_t* payload = (uint32_t*)dp.data();
+
+            for (uint32_t i = 0; i < 2; i++) {
+                payload[i] = ntohl(payload[i]);
+            }
+            dprintf("Read addr[%x] = %x \n", payload[0], payload[1]);
```

Comment:
Assuming these prints will be removed or changed to status message that can be emitted to the GUI before committing, so will not comment on the remainder of them.

Reply:
Removed.